File: tools/quantize_int4.mjs
#!/usr/bin/env node
// llm-on-cpu :: tools/quantize_int4.mjs
// BF16/F16 LWC → QLWC INT4 —— quantize_int4.py 的 Node 对等实现（零 npm 依赖）。
// 布局与 C++ src/weights/qlwc_format.* / Python 版字节级一致。
//
// 用法:
//   node tools/quantize_int4.mjs --src models/Qwen3.5-4B.lwc --out models/Qwen3.5-4B.int4.qlwc --method awq
//   node tools/quantize_int4.mjs --src ... --out ... --method gptq --group-size 128

import fs from "node:fs";
import path from "node:path";

const ALIGN = 4096;
const QLW_MAGIC = Buffer.from("QLW1");
const SCHEME_GPTQ = 1;
const SCHEME_AWQ = 2;

function fnv1a64(buf) {
  let h = 0xcbf29ce484222325n;
  const P = 0x100000001b3n;
  const M = (1n << 64n) - 1n;
  for (const b of buf) h = ((h ^ BigInt(b)) * P) & M;
  return h;
}

function alignUp(v, a) {
  return Math.floor((v + a - 1) / a) * a;
}

function parseArgs() {
  const a = process.argv.slice(2);
  const get = (k) => {
    const i = a.indexOf(k);
    return i >= 0 ? a[i + 1] : undefined;
  };
  if (!a.includes("--src") || !a.includes("--out")) {
    console.error(
      "usage: node quantize_int4.mjs --src <file.lwc> --out <file.qlwc> [--method awq|gptq] [--group-size 128] [--min-cols 128]",
    );
    process.exit(2);
  }
  const method = get("--method") ?? "awq";
  if (method !== "awq" && method !== "gptq") {
    console.error("--method must be awq or gptq");
    process.exit(2);
  }
  return {
    src: get("--src"),
    out: get("--out"),
    method,
    groupSize: Number(get("--group-size") ?? 128),
    minCols: Number(get("--min-cols") ?? 128),
  };
}

function u32(v) {
  const b = Buffer.alloc(4);
  b.writeUInt32LE(v >>> 0);
  return b;
}
function u64(v) {
  const b = Buffer.alloc(8);
  b.writeBigUInt64LE(BigInt(v));
  return b;
}
function str(s) {
  const b = Buffer.from(s, "utf8");
  return Buffer.concat([u32(b.length), b]);
}

/** IEEE754 f32 → f16 bits (uint16) */
function f32ToF16Bits(x) {
  const f32 = new Float32Array(1);
  const u32v = new Uint32Array(f32.buffer);
  f32[0] = x;
  const bits = u32v[0];
  const sign = (bits >>> 16) & 0x8000;
  let exp = (bits >>> 23) & 0xff;
  let mant = bits & 0x7fffff;
  if (exp === 255) {
    // Inf/NaN
    return sign | 0x7c00 | (mant ? 0x200 : 0);
  }
  if (exp === 0) {
    // subnormal/zero → flush to 0 in f16
    return sign;
  }
  exp = exp - 127 + 15;
  if (exp >= 31) return sign | 0x7c00; // overflow → Inf
  if (exp <= 0) {
    if (exp < -10) return sign;
    mant |= 0x800000;
    const shift = 14 - exp;
    let half = mant >>> (shift + 13);
    // round
    if ((mant >>> (shift + 12)) & 1) half += 1;
    return sign | half;
  }
  let half = (exp << 10) | (mant >>> 13);
  if (mant & 0x1000) half += 1; // round
  return sign | (half & 0x7fff);
}

function f16BitsToF32(h) {
  const sign = (h & 0x8000) << 16;
  let exp = (h >>> 10) & 0x1f;
  let mant = h & 0x3ff;
  let bits;
  if (exp === 0) {
    if (mant === 0) bits = sign;
    else {
      exp = 1;
      while (!(mant & 0x400)) {
        mant <<= 1;
        exp--;
      }
      mant &= 0x3ff;
      bits = sign | ((exp + 127 - 15) << 23) | (mant << 13);
    }
  } else if (exp === 31) {
    bits = sign | 0x7f800000 | (mant << 13);
  } else {
    bits = sign | ((exp + 127 - 15) << 23) | (mant << 13);
  }
  const u = new Uint32Array([bits]);
  return new Float32Array(u.buffer)[0];
}

function bf16BytesToF32(buf, n) {
  const u16 = new Uint16Array(buf.buffer, buf.byteOffset, n);
  const out = new Float32Array(n);
  const tmp = new Uint32Array(1);
  const fview = new Float32Array(tmp.buffer);
  for (let i = 0; i < n; ++i) {
    tmp[0] = u16[i] << 16;
    out[i] = fview[0];
  }
  return out;
}

function f16BytesToF32(buf, n) {
  const u16 = new Uint16Array(buf.buffer, buf.byteOffset, n);
  const out = new Float32Array(n);
  for (let i = 0; i < n; ++i) out[i] = f16BitsToF32(u16[i]);
  return out;
}

function f32ArrToF16Bits(arr) {
  const out = new Uint16Array(arr.length);
  for (let i = 0; i < arr.length; ++i) out[i] = f32ToF16Bits(arr[i]);
  return out;
}

function readLwc(filePath) {
  const fd = fs.openSync(filePath, "r");
  try {
    const magic = Buffer.alloc(4);
    fs.readSync(fd, magic, 0, 4, 0);
    if (magic.toString("ascii") !== "LWC1") throw new Error(`not LWC1: ${filePath}`);
    const pref = Buffer.alloc(20);
    fs.readSync(fd, pref, 0, 20, 4);
    const ver = pref.readUInt32LE(0);
    const catLen = Number(pref.readBigUInt64LE(4));
    const catCrc = pref.readBigUInt64LE(12);
    void ver;
    const cat = Buffer.alloc(catLen);
    fs.readSync(fd, cat, 0, catLen, 24);
    if (fnv1a64(cat) !== catCrc) throw new Error("LWC catalog crc mismatch");

    let o = 0;
    const u32r = () => {
      const v = cat.readUInt32LE(o);
      o += 4;
      return v;
    };
    const u64r = () => {
      const v = Number(cat.readBigUInt64LE(o));
      o += 8;
      return v;
    };
    const sr = () => {
      const n = u32r();
      const s = cat.subarray(o, o + n).toString("utf8");
      o += n;
      return s;
    };

    const dtype = u32r();
    u32r(); // align
    const nT = u64r();
    const nG = u64r();
    const tensors = [];
    for (let i = 0; i < nT; ++i) {
      const name = sr();
      const nd = u64r();
      const shape = [];
      for (let d = 0; d < nd; ++d) shape.push(u64r());
      const offset = u64r();
      const nbytes = u64r();
      u64r(); // checksum
      tensors.push({ name, shape, offset, nbytes });
    }
    for (let g = 0; g < nG; ++g) {
      u32r();
      u32r();
      const nn = u64r();
      for (let i = 0; i < nn; ++i) sr();
    }

    const payloads = {};
    for (const t of tensors) {
      const buf = Buffer.alloc(t.nbytes);
      fs.readSync(fd, buf, 0, t.nbytes, t.offset);
      payloads[t.name] = buf;
    }
    return { dtype, tensors, payloads };
  } finally {
    fs.closeSync(fd);
  }
}

function packInt4(q /* Uint8Array flat M*K */, M, K) {
  let Kp = K;
  let src = q;
  if (K % 2) {
    Kp = K + 1;
    src = new Uint8Array(M * Kp);
    for (let m = 0; m < M; ++m) {
      src.set(q.subarray(m * K, m * K + K), m * Kp);
    }
  }
  const packed = new Uint8Array(M * (Kp / 2));
  for (let m = 0; m < M; ++m) {
    const row = m * Kp;
    const prow = m * (Kp / 2);
    for (let k = 0; k < Kp; k += 2) {
      packed[prow + k / 2] = (src[row + k] & 0xf) | ((src[row + k + 1] & 0xf) << 4);
    }
  }
  return Buffer.from(packed.buffer, packed.byteOffset, packed.byteLength);
}

function quantGptqAsym(W /* Float32Array */, M, K, groupSize) {
  if (K % groupSize !== 0) throw new Error("K % group_size != 0");
  const ng = K / groupSize;
  const scales = new Float32Array(M * ng);
  const zeros = new Float32Array(M * ng);
  const q = new Uint8Array(M * K);
  for (let g = 0; g < ng; ++g) {
    const k0 = g * groupSize;
    for (let m = 0; m < M; ++m) {
      let mn = Infinity;
      let mx = -Infinity;
      const base = m * K + k0;
      for (let k = 0; k < groupSize; ++k) {
        const v = W[base + k];
        if (v < mn) mn = v;
        if (v > mx) mx = v;
      }
      let scale = (mx - mn) / 15.0;
      if (scale < 1e-8) scale = 1e-8;
      scales[m * ng + g] = scale;
      zeros[m * ng + g] = mn;
      for (let k = 0; k < groupSize; ++k) {
        let qq = Math.round((W[base + k] - mn) / scale);
        if (qq < 0) qq = 0;
        if (qq > 15) qq = 15;
        q[base + k] = qq;
      }
    }
  }
  return {
    q: packInt4(q, M, K),
    scales: f32ArrToF16Bits(scales),
    zeros: f32ArrToF16Bits(zeros),
  };
}

function quantAwqSym(W, M, K, groupSize) {
  if (K % groupSize !== 0) throw new Error("K % group_size != 0");
  const ng = K / groupSize;
  // 对称 absmax（AWQ 兼容无-zero 布局；非完整 AWQ 校准）
  const scales = new Float32Array(M * ng);
  const q = new Uint8Array(M * K);
  for (let g = 0; g < ng; ++g) {
    const k0 = g * groupSize;
    for (let m = 0; m < M; ++m) {
      const base = m * K + k0;
      let amax = 0;
      for (let k = 0; k < groupSize; ++k) {
        const a = Math.abs(W[base + k]);
        if (a > amax) amax = a;
      }
      let scale = amax / 7.0;
      if (scale < 1e-8) scale = 1e-8;
      scales[m * ng + g] = scale;
      for (let k = 0; k < groupSize; ++k) {
        let qq = Math.round(W[base + k] / scale) + 7;
        if (qq < 0) qq = 0;
        if (qq > 15) qq = 15;
        q[base + k] = qq;
      }
    }
  }
  return { q: packInt4(q, M, K), scales: f32ArrToF16Bits(scales), zeros: null };
}

function writeAt(fd, buf, filePos) {
  // Node writeSync length 上限为 2^31-1；大文件必须分块写
  const CHUNK = 1 << 30; // 1 GiB
  let off = 0;
  let pos = filePos;
  while (off < buf.length) {
    const n = Math.min(CHUNK, buf.length - off);
    const w = fs.writeSync(fd, buf, off, n, pos);
    off += w;
    pos += w;
  }
}

function writeQlwc(filePath, scheme, groupSize, items) {
  const buildCat = (withOff, offsets) => {
    const parts = [u32(scheme), u32(groupSize), u32(ALIGN), u64(items.length)];
    for (let i = 0; i < items.length; ++i) {
      const it = items[i];
      parts.push(str(it.name), u32(it.kind), u64(it.shape.length));
      for (const d of it.shape) parts.push(u64(d));
      if (it.kind === 0) {
        parts.push(u32(it.pass_dtype));
        const off = withOff ? offsets[i].data : 0;
        parts.push(u64(off), u64(it.data.length));
      } else {
        const o = withOff ? offsets[i] : { q: 0, s: 0, z: 0 };
        const zlen = it.zeros ? it.zeros.length * 2 : 0;
        parts.push(
          u32(groupSize),
          u64(o.q),
          u64(it.q.length),
          u64(o.s),
          u64(it.scales.length * 2),
          u64(o.z),
          u64(zlen),
        );
      }
    }
    return Buffer.concat(parts);
  };

  const probe = buildCat(false, null);
  let off = alignUp(24 + probe.length, ALIGN);
  const offsets = [];
  for (const it of items) {
    if (it.kind === 0) {
      offsets.push({ data: off });
      off = alignUp(off + it.data.length, ALIGN);
    } else {
      const qOff = off;
      off = alignUp(off + it.q.length, 64);
      const sOff = off;
      off = alignUp(off + it.scales.length * 2, 64);
      const zOff = off;
      const zlen = it.zeros ? it.zeros.length * 2 : 0;
      off = alignUp(off + zlen, ALIGN);
      offsets.push({ q: qOff, s: sOff, z: zOff });
    }
  }

  const catalog = buildCat(true, offsets);
  const crc = fnv1a64(catalog);
  const preface = Buffer.concat([
    QLW_MAGIC,
    u32(1),
    u64(catalog.length),
    (() => {
      const b = Buffer.alloc(8);
      b.writeBigUInt64LE(crc);
      return b;
    })(),
  ]);
  const data0 = alignUp(preface.length + catalog.length, ALIGN);
  const fileSize = Math.max(off, data0);

  fs.mkdirSync(path.dirname(filePath), { recursive: true });
  const fd = fs.openSync(filePath, "w");
  try {
    // 不整文件 Buffer.alloc：>2GiB 会撞 Node write 长度上限
    writeAt(fd, preface, 0);
    writeAt(fd, catalog, preface.length);
    fs.ftruncateSync(fd, fileSize); // 对齐空洞填 0
    for (let i = 0; i < items.length; ++i) {
      const it = items[i];
      if (it.kind === 0) {
        writeAt(fd, it.data, offsets[i].data);
      } else {
        const o = offsets[i];
        writeAt(fd, it.q, o.q);
        writeAt(
          fd,
          Buffer.from(it.scales.buffer, it.scales.byteOffset, it.scales.byteLength),
          o.s,
        );
        if (it.zeros) {
          writeAt(
            fd,
            Buffer.from(it.zeros.buffer, it.zeros.byteOffset, it.zeros.byteLength),
            o.z,
          );
        }
      }
      if ((i + 1) % 50 === 0)
        process.stderr.write(`\r[quantize_int4] writing ${i + 1}/${items.length}...`);
    }
    if (items.length >= 50) process.stderr.write("\n");
  } finally {
    fs.closeSync(fd);
  }
}

function shouldQuantize(name, shape, minCols, gs) {
  if (shape.length !== 2 || shape[1] < minCols || shape[1] % gs !== 0) return false;
  const n = name.toLowerCase();
  if (n.endsWith("embed_tokens.weight") || n.endsWith("embedding.weight")) return false;
  if (n.endsWith("lm_head.weight") || n === "lm_head.weight") return false;
  return true;
}

function main() {
  const opt = parseArgs();
  const src = path.resolve(opt.src);
  const out = path.resolve(opt.out);
  const { dtype, tensors, payloads } = readLwc(src);
  const toF32 = dtype === 1 ? bf16BytesToF32 : f16BytesToF32;
  const scheme = opt.method === "awq" ? SCHEME_AWQ : SCHEME_GPTQ;
  const gs = opt.groupSize;
  const items = [];
  let nQ = 0;
  let nPass = 0;

  for (const t of tensors) {
    const { name, shape } = t;
    const raw = payloads[name];
    const ne = shape.reduce((a, b) => a * b, 1);
    if (shouldQuantize(name, shape, opt.minCols, gs)) {
      const M = shape[0];
      const K = shape[1];
      const W = toF32(raw, ne);
      const qres =
        opt.method === "awq" ? quantAwqSym(W, M, K, gs) : quantGptqAsym(W, M, K, gs);
      items.push({
        name,
        kind: 1,
        shape,
        q: qres.q,
        scales: qres.scales,
        zeros: qres.zeros,
      });
      nQ += 1;
      if (nQ % 20 === 0) process.stderr.write(`\r[quantize_int4] quantized ${nQ} tensors...`);
    } else {
      items.push({ name, kind: 0, shape, pass_dtype: dtype, data: raw });
      nPass += 1;
    }
  }
  if (nQ >= 20) process.stderr.write("\n");

  writeQlwc(out, scheme, gs, items);
  const sizeG = fs.statSync(out).size / 1024 ** 3;
  console.log(`[quantize_int4] method=${opt.method} group=${gs}`);
  console.log(`[quantize_int4] quantized=${nQ} passthrough=${nPass}`);
  console.log(`[quantize_int4] wrote ${out} (${sizeG.toFixed(2)} GiB)`);
  console.log(`[next] .\\build\\msvc-x64\\bin\\llmoc_server_int4.exe --config configs/engine_int4.yaml`);
}

main();

File: tools/quantize_int4.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
quantize_int4.py — BF16/F16 LWC → QLWC INT4（GPTQ/AWQ 风格布局，独立新功能）

不修改原 convert_lwc / LWC 推理路径。

用法:
  python tools/quantize_int4.py --src models/Qwen3.5-4B.lwc --out models/Qwen3.5-4B.int4.qlwc --method awq
  node   tools/quantize_int4.mjs --src models/Qwen3.5-4B.lwc --out models/Qwen3.5-4B.int4.qlwc --method awq
  python tools/quantize_int4.py --src models/Qwen3.5-4B.lwc --out models/Qwen3.5-4B.int4.qlwc --method gptq --group-size 128

与 convert_lwc 一样提供 Python / Node 双栈（Node 零 npm 依赖；算法与布局对齐）。

方法说明:
  awq  — 对称 INT4 absmax，AWQ 兼容存储（无 zero）；无校准集，非完整 AWQ
  gptq — 非对称 INT4 per-group scale+zero；布局对齐 GPTQ，无 Hessian 校准

embedding / lm_head 默认透传原精度（切勿量化，否则又慢又胡话）。
"""
from __future__ import annotations

import argparse
import struct
from pathlib import Path

import numpy as np

ALIGN = 4096
QLW_MAGIC = b"QLW1"
SCHEME_GPTQ = 1
SCHEME_AWQ = 2


def fnv1a64(data: bytes) -> int:
    h = 14695981039346656037
    for b in data:
        h = ((h ^ b) * 1099511628211) & ((1 << 64) - 1)
    return h


def align_up(v: int, a: int) -> int:
    return (v + a - 1) // a * a


def f32_to_f16_bits(x: np.ndarray) -> np.ndarray:
    return x.astype(np.float16).view(np.uint16)


def bf16_bytes_to_f32(buf: bytes, n: int) -> np.ndarray:
    u16 = np.frombuffer(buf, dtype=np.uint16, count=n)
    u32 = u16.astype(np.uint32) << 16
    return u32.view(np.float32)


def f16_bytes_to_f32(buf: bytes, n: int) -> np.ndarray:
    return np.frombuffer(buf, dtype=np.float16, count=n).astype(np.float32)


def read_lwc(path: Path):
    with path.open("rb") as f:
        magic = f.read(4)
        if magic != b"LWC1":
            raise SystemExit(f"not LWC1: {path}")
        ver, cat_len, cat_crc = struct.unpack("<IQQ", f.read(20))
        cat = f.read(cat_len)
        if fnv1a64(cat) != cat_crc:
            raise SystemExit("LWC catalog crc mismatch")
        raw = memoryview(cat)
        o = 0

        def u32():
            nonlocal o
            v = struct.unpack_from("<I", raw, o)[0]
            o += 4
            return v

        def u64():
            nonlocal o
            v = struct.unpack_from("<Q", raw, o)[0]
            o += 8
            return v

        def s():
            nonlocal o
            n = u32()
            out = bytes(raw[o : o + n]).decode("utf-8")
            o += n
            return out

        dtype = u32()
        align = u32()
        n_t, n_g = u64(), u64()
        tensors = []
        for _ in range(n_t):
            name = s()
            nd = u64()
            shape = [u64() for _ in range(nd)]
            off, nb, ck = u64(), u64(), u64()
            tensors.append({"name": name, "shape": shape, "offset": off, "nbytes": nb})
        # skip groups
        for _ in range(n_g):
            u32()
            u32()
            nn = u64()
            for _ in range(nn):
                s()
        payloads = {}
        for t in tensors:
            f.seek(t["offset"])
            payloads[t["name"]] = f.read(t["nbytes"])
    return dtype, tensors, payloads


def pack_int4(q: np.ndarray) -> bytes:
    """q: uint8 0..15, shape [M,K] even K preferred."""
    M, K = q.shape
    if K % 2:
        q = np.pad(q, ((0, 0), (0, 1)), constant_values=0)
        K += 1
    lo = q[:, 0::2]
    hi = q[:, 1::2]
    packed = (lo | (hi << 4)).astype(np.uint8)
    return packed.tobytes()


def quant_gptq_asym(W: np.ndarray, group_size: int):
    """W [M,K] f32 → q uint8, scales/zeros f16 bits."""
    M, K = W.shape
    assert K % group_size == 0
    ng = K // group_size
    scales = np.zeros((M, ng), np.float32)
    zeros = np.zeros((M, ng), np.float32)
    q = np.zeros((M, K), np.uint8)
    for g in range(ng):
        sl = slice(g * group_size, (g + 1) * group_size)
        block = W[:, sl]
        mn = block.min(axis=1)
        mx = block.max(axis=1)
        scale = (mx - mn) / 15.0
        scale = np.maximum(scale, 1e-8)
        zero = mn
        scales[:, g] = scale
        zeros[:, g] = zero
        qq = np.round((block - zero[:, None]) / scale[:, None])
        q[:, sl] = np.clip(qq, 0, 15).astype(np.uint8)
    return pack_int4(q), f32_to_f16_bits(scales.reshape(-1)), f32_to_f16_bits(zeros.reshape(-1))


def quant_awq_sym(W: np.ndarray, group_size: int):
    """对称 INT4（AWQ 兼容存储：无 zero）。无校准集时用 per-group absmax，非完整 AWQ。"""
    M, K = W.shape
    assert K % group_size == 0
    ng = K // group_size
    scales = np.zeros((M, ng), np.float32)
    q = np.zeros((M, K), np.uint8)
    for g in range(ng):
        sl = slice(g * group_size, (g + 1) * group_size)
        block = W[:, sl]
        amax = np.max(np.abs(block), axis=1)
        scale = np.maximum(amax / 7.0, 1e-8)
        scales[:, g] = scale
        qq = np.round(block / scale[:, None]) + 7
        q[:, sl] = np.clip(qq, 0, 15).astype(np.uint8)
    # dequant: (q-7)*scale ≈ W
    return pack_int4(q), f32_to_f16_bits(scales.reshape(-1)), None


def put_u32(buf: bytearray, v: int):
    buf.extend(struct.pack("<I", v))


def put_u64(buf: bytearray, v: int):
    buf.extend(struct.pack("<Q", v))


def put_str(buf: bytearray, s: str):
    b = s.encode("utf-8")
    put_u32(buf, len(b))
    buf.extend(b)


def write_qlwc(path: Path, scheme: int, group_size: int, items: list):
    # items: dict name, kind, shape, pass_dtype?, data?, q?, scales?, zeros?
    cat = bytearray()
    put_u32(cat, scheme)
    put_u32(cat, group_size)
    put_u32(cat, ALIGN)
    put_u64(cat, len(items))
    # placeholder offsets — build twice
    def build_cat(with_off: bool, offsets: list | None = None):
        c = bytearray()
        put_u32(c, scheme)
        put_u32(c, group_size)
        put_u32(c, ALIGN)
        put_u64(c, len(items))
        for i, it in enumerate(items):
            put_str(c, it["name"])
            put_u32(c, it["kind"])
            put_u64(c, len(it["shape"]))
            for d in it["shape"]:
                put_u64(c, int(d))
            if it["kind"] == 0:
                put_u32(c, it["pass_dtype"])
                off = offsets[i]["data"] if with_off else 0
                put_u64(c, off)
                put_u64(c, len(it["data"]))
            else:
                put_u32(c, group_size)
                o = offsets[i] if with_off else {"q": 0, "s": 0, "z": 0}
                put_u64(c, o["q"])
                put_u64(c, len(it["q"]))
                put_u64(c, o["s"])
                put_u64(c, len(it["scales"]) * 2)
                put_u64(c, o["z"])
                zlen = 0 if it["zeros"] is None else len(it["zeros"]) * 2
                put_u64(c, zlen)
        return c

    probe = build_cat(False)
    off = align_up(24 + len(probe), ALIGN)
    offsets = []
    for it in items:
        if it["kind"] == 0:
            offsets.append({"data": off})
            off = align_up(off + len(it["data"]), ALIGN)
        else:
            q_off = off
            off = align_up(off + len(it["q"]), 64)
            s_off = off
            off = align_up(off + len(it["scales"]) * 2, 64)
            z_off = off
            zlen = 0 if it["zeros"] is None else len(it["zeros"]) * 2
            off = align_up(off + zlen, ALIGN)
            offsets.append({"q": q_off, "s": s_off, "z": z_off})

    catalog = build_cat(True, offsets)
    preface = QLW_MAGIC + struct.pack("<IQQ", 1, len(catalog), fnv1a64(bytes(catalog)))
    data0 = align_up(len(preface) + len(catalog), ALIGN)
    path.parent.mkdir(parents=True, exist_ok=True)
    # 流式按偏移写盘，避免整文件 >2GiB 时双倍内存
    with path.open("wb") as f:
        f.write(preface)
        f.write(catalog)
        if f.tell() < data0:
            f.write(b"\0" * (data0 - f.tell()))
        f.truncate(off)
        for i, it in enumerate(items):
            if it["kind"] == 0:
                f.seek(offsets[i]["data"])
                f.write(it["data"])
            else:
                o = offsets[i]
                f.seek(o["q"])
                f.write(it["q"])
                f.seek(o["s"])
                f.write(it["scales"].astype(np.uint16).tobytes())
                if it["zeros"] is not None:
                    f.seek(o["z"])
                    f.write(it["zeros"].astype(np.uint16).tobytes())


def should_quantize(name: str, shape: list, min_cols: int, gs: int) -> bool:
    if len(shape) != 2 or shape[1] < min_cols or shape[1] % gs != 0:
        return False
    n = name.lower()
    # embedding / lm_head 保持原精度（查表 + 大 vocab GEMM）
    if n.endswith("embed_tokens.weight") or n.endswith("embedding.weight"):
        return False
    if n.endswith("lm_head.weight") or n == "lm_head.weight":
        return False
    return True


def main() -> int:
    ap = argparse.ArgumentParser(description="BF16/F16 LWC → QLWC INT4 (GPTQ/AWQ layout)")
    ap.add_argument("--src", required=True, help="输入 .lwc (BF16/F16)")
    ap.add_argument("--out", required=True, help="输出 .qlwc")
    ap.add_argument("--method", choices=("awq", "gptq"), default="awq")
    ap.add_argument("--group-size", type=int, default=128)
    ap.add_argument("--min-cols", type=int, default=128, help="列数低于此不量化(透传)")
    args = ap.parse_args()

    dtype, tensors, payloads = read_lwc(Path(args.src))
    to_f32 = bf16_bytes_to_f32 if dtype == 1 else f16_bytes_to_f32
    scheme = SCHEME_AWQ if args.method == "awq" else SCHEME_GPTQ
    gs = args.group_size
    items = []
    n_q, n_pass = 0, 0
    for t in tensors:
        name, shape, raw = t["name"], t["shape"], payloads[t["name"]]
        ne = int(np.prod(shape)) if shape else 0
        # 仅 2D 权重且 K 对齐 group；embedding/lm_head 透传
        if should_quantize(name, shape, args.min_cols, gs):
            W = to_f32(raw, ne).reshape(int(shape[0]), int(shape[1]))
            if args.method == "awq":
                q, scales, zeros = quant_awq_sym(W, gs)
            else:
                q, scales, zeros = quant_gptq_asym(W, gs)
            items.append(
                {
                    "name": name,
                    "kind": 1,
                    "shape": shape,
                    "q": q,
                    "scales": scales,
                    "zeros": zeros,
                }
            )
            n_q += 1
        else:
            items.append(
                {
                    "name": name,
                    "kind": 0,
                    "shape": shape,
                    "pass_dtype": dtype,
                    "data": raw,
                }
            )
            n_pass += 1

    out = Path(args.out)
    write_qlwc(out, scheme, gs, items)
    size_g = out.stat().st_size / (1024**3)
    print(f"[quantize_int4] method={args.method} group={gs}")
    print(f"[quantize_int4] quantized={n_q} passthrough={n_pass}")
    print(f"[quantize_int4] wrote {out} ({size_g:.2f} GiB)")
    print(f"[next] .\\build\\msvc-x64\\bin\\llmoc_server_int4.exe --config configs/engine_int4.yaml")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
